python/ray/util/sgd/torch/examples/gat_dgl.py

```diff
@@ -0,0 +1,274 @@
+import os
+import torch
+import time
+import numpy as np
+import argparse
+from dgl.dataloading import NodeCollator
+import ray
+from ray.util.sgd import TorchTrainer
+from ray.util.sgd.utils import AverageMeterCollection
+from ray.util.sgd.torch import TrainingOperator
+import dgl
+import torch as th
+import torch.nn as nn
+import torch.nn.functional as F
+from dgl.data import RedditDataset
+from dgl.nn.pytorch import GATConv
+from torch.utils.data import DataLoader
+
+print("Current Path: " + os.getcwd())
+torch.manual_seed(42)
+
+
+# define the model class
+class GAT(nn.Module):
+    def __init__(self, in_feats, n_hidden, n_classes, n_layers, n_heads,
+                 activation, feat_drop, attn_drop, negative_slope, residual):
+        super().__init__()
+
+        self.n_layers = n_layers
+        self.activation = activation
+        self.n_hidden = n_hidden
+        self.n_heads = n_heads
+        self.n_classes = n_classes
+        self.convs = nn.ModuleList()
+
+        # input layer
+        self.convs.append(
+            GATConv((in_feats, in_feats), n_hidden, n_heads, feat_drop,
+                    attn_drop, negative_slope, residual, self.activation))
+        # hidden layer
+        for l in range(1, n_layers - 1):
+            # due to multi-head, the in_dim = num_hidden * num_heads
+            self.convs.append(
+                GATConv((n_hidden * n_heads, n_hidden * n_heads), n_hidden,
+                        n_heads, feat_drop, attn_drop, negative_slope,
+                        residual, self.activation))
+        ##output layer
+        self.convs.append(
+            GATConv((n_hidden * n_heads, n_hidden * n_heads), n_classes,
+                    n_heads, feat_drop, attn_drop, negative_slope, residual,
+                    None))
+
+    def forward(self, blocks, x):
+        h = x
+        for l, (layer, block) in enumerate(zip(self.convs, blocks)):
+            h_dst = h[:block.number_of_dst_nodes()]
+            if l != len(self.convs) - 1:
+                h = layer(block, (h, h_dst)).flatten(1)
+                h = F.dropout(h, p=0.5, training=self.training)
+            else:
+                h = layer(block, (h, h_dst))
+        h = h.mean(1)
+        return h.log_softmax(dim=-1)
+
+
+def compute_acc(pred, labels):
+    """
+    Compute the accuracy of prediction given the labels.
+    """
+    return (th.argmax(pred, dim=1) == labels).float().sum() / len(pred)
+
+
+class _NodeCollator(NodeCollator):
```

Comment:
I might be missing something here, but is this necessary? It looks like it's doing the same thing as the superclass

Reply:
Thank you for your suggestion, the code has been deleted, and the super class(NodeCollator) is used directly to load data in the custom class(CustomTrainingOperator)